File: apps/control-plane/src/sessions/manager.ts
/**
 * Session Manager
 *
 * Manages Claude Code agent sessions using Kubernetes and agent-sandbox.
 * Persists session state, messages, and events to Redis.
 */
import type {
  AgentEvent,
  Session,
  SessionCreateRequest,
  ServerMessage,
  PersistedMessage,
  PersistedEvent,
  SessionWithMeta,
} from "@netclode/protocol";
import { KubernetesRuntime } from "../runtime/kubernetes";
import { config } from "../config";
import type { RedisStorage } from "../storage/redis";

// Kubernetes runtime instance
const runtime = new KubernetesRuntime();

export type MessageHandler = (message: ServerMessage) => void;

interface AgentConnection {
  serviceFQDN: string;
  connected: boolean;
  close: () => void;
}

interface SessionState {
  session: Session;
  agent: AgentConnection | null;
  messageHandlers: Set<MessageHandler>;
}

export class SessionManager {
  private sessions: Map<string, SessionState> = new Map();
  private storage: RedisStorage;
  private initialized = false;

  constructor(storage: RedisStorage) {
    this.storage = storage;
  }

  /**
   * Initialize the session manager by loading sessions from Redis
   * and reconciling with K8s state.
   */
  async initialize(): Promise<void> {
    if (this.initialized) return;

    console.log("[sessions] Loading sessions from Redis...");
    await this.loadFromRedis();

    console.log("[sessions] Reconciling with K8s state...");
    await this.reconcileWithK8s();

    this.initialized = true;
    console.log(`[sessions] Initialized with ${this.sessions.size} sessions`);
  }

  private async loadFromRedis(): Promise<void> {
    const sessions = await this.storage.getAllSessions();
    for (const session of sessions) {
      this.sessions.set(session.id, {
        session,
        agent: null,
        messageHandlers: new Set(),
      });
    }
  }

  private async reconcileWithK8s(): Promise<void> {
    const sandboxes = await runtime.listSandboxes();
    const sandboxMap = new Map(
      sandboxes.map((s) => [s.name.replace("sess-", ""), s])
    );

    for (const [id, state] of this.sessions) {
      const sandbox = sandboxMap.get(id);

      if (!sandbox) {
        // No sandbox - mark as paused if currently running
        if (
          state.session.status === "running" ||
          state.session.status === "creating"
        ) {
          state.session.status = "paused";
          await this.storage.updateSessionField(id, "status", "paused");
          console.log(`[${id}] Reconciled: no sandbox, marked paused`);
        }
      } else if (sandbox.status === "ready") {
        // Sandbox ready - could reconnect
        if (state.session.status === "paused") {
          // Don't auto-resume, just note it's available
          console.log(`[${id}] Reconciled: sandbox ready, session paused`);
        }
      }
    }
  }

  async create(request: SessionCreateRequest): Promise<Session> {
    const id = crypto.randomUUID().slice(0, 12);
    const session: Session = {
      id,
      name: request.name || `session-${id.slice(0, 6)}`,
      status: "creating",
      repo: request.repo,
      createdAt: new Date().toISOString(),
      lastActiveAt: new Date().toISOString(),
    };

    // Persist to Redis first
    await this.storage.saveSession(session);

    this.sessions.set(id, {
      session,
      agent: null,
      messageHandlers: new Set(),
    });

    this.startSessionCreation(id, request).catch((error) => {
      console.error(`[${id}] Failed to create session:`, error);
    });

    return session;
  }

  async list(): Promise<Session[]> {
    // Sync with actual sandbox state
    const sandboxes = await runtime.listSandboxes();
    const sandboxIds = new Set(
      sandboxes.map((s) => s.name.replace("sess-", ""))
    );

    // Update statuses based on actual sandbox state
    for (const [id, state] of this.sessions) {
      if (!sandboxIds.has(id) && state.session.status === "running") {
        state.session.status = "paused";
        // Persist status change
        await this.storage.updateSessionField(id, "status", "paused");
      }
    }

    return Array.from(this.sessions.values()).map((s) => s.session);
  }

  async get(id: string): Promise<Session | undefined> {
    return this.sessions.get(id)?.session;
  }

  async resume(id: string): Promise<Session> {
    const state = this.sessions.get(id);
    if (!state) {
      throw new Error(`Session ${id} not found`);
    }

    // Check if sandbox is running
    let sandboxInfo = await runtime.getSandboxStatus(id);
    let serviceFQDN = sandboxInfo?.serviceFQDN;

    if (
      state.session.status === "creating" &&
      sandboxInfo?.status !== "ready"
    ) {
      serviceFQDN = (await runtime.waitForReady(id, 120000)) ?? undefined;
      if (!serviceFQDN) {
        state.session.status = "error";
        await this.storage.updateSessionField(id, "status", "error");
        this.emitSession(state, {
          type: "session.updated",
          session: state.session,
        });
        throw new Error("Failed to resume session");
      }
    } else if (!serviceFQDN || sandboxInfo?.status !== "ready") {
      // Recreate sandbox
      console.log(`[${id}] Sandbox not running, recreating...`);

      // Delete old sandbox if exists
      if (sandboxInfo) {
        await runtime.deleteSandbox(id);
      }

      await runtime.createSandbox({
        sessionId: id,
        cpus: config.defaultCpus,
        memoryMB: config.defaultMemoryMB,
      });

      serviceFQDN = (await runtime.waitForReady(id, 120000)) ?? undefined;
      if (!serviceFQDN) {
        state.session.status = "error";
        await this.storage.updateSessionField(id, "status", "error");
        this.emitSession(state, {
          type: "session.updated",
          session: state.session,
        });
        throw new Error("Failed to resume session");
      }
    }

    state.session.status = "running";
    state.session.lastActiveAt = new Date().toISOString();

    // Persist status change
    await this.storage.updateSessionFields(id, {
      status: "running",
      lastActiveAt: state.session.lastActiveAt,
    });

    // Store agent connection
    state.agent = {
      serviceFQDN,
      connected: true,
      close: () => {
        if (state.agent) state.agent.connected = false;
      },
    };

    console.log(`[${id}] Resumed, agent at ${serviceFQDN}`);
    return state.session;
  }

  async pause(id: string): Promise<Session> {
    const state = this.sessions.get(id);
    if (!state) {
      throw new Error(`Session ${id} not found`);
    }

    // Disconnect agent
    if (state.agent) {
      state.agent.close();
      state.agent = null;
    }

    // Delete sandbox (PVC persists)
    await runtime.deleteSandbox(id);

    state.session.status = "paused";
    state.session.lastActiveAt = new Date().toISOString();

    // Persist status change
    await this.storage.updateSessionFields(id, {
      status: "paused",
      lastActiveAt: state.session.lastActiveAt,
    });

    return state.session;
  }

  async delete(id: string): Promise<void> {
    const state = this.sessions.get(id);
    if (!state) {
      throw new Error(`Session ${id} not found`);
    }

    // Disconnect agent
    if (state.agent) {
      state.agent.close();
    }

    // Delete sandbox and associated resources
    await runtime.deleteSandbox(id);

    // Remove from Redis
    await this.storage.deleteSession(id);

    this.sessions.delete(id);
  }

  /**
   * Send a prompt to the agent and stream responses.
   * Persists user message and final assistant response to Redis.
   */
  async sendPrompt(id: string, text: string): Promise<void> {
    console.log(`[${id}] sendPrompt called: "${text.slice(0, 50)}..."`);

    const state = this.sessions.get(id);
    if (!state) {
      console.error(`[${id}] sendPrompt: Session not found in memory`);
      throw new Error(`Session ${id} not found`);
    }

    console.log(`[${id}] sendPrompt: agent connected=${state.agent?.connected}, fqdn=${state.agent?.serviceFQDN}`);
    if (!state.agent?.connected || !state.agent.serviceFQDN) {
      console.error(`[${id}] sendPrompt: Agent not connected`);
      throw new Error(`Session ${id} agent not connected`);
    }

    state.session.lastActiveAt = new Date().toISOString();
    state.session.status = "running";

    // Persist user message
    const userMessageId = `msg_${crypto.randomUUID().slice(0, 12)}`;
    console.log(`[${id}] sendPrompt: Persisting user message ${userMessageId}`);
    await this.storage.appendMessage(id, {
      id: userMessageId,
      sessionId: id,
      role: "user",
      content: text,
      timestamp: new Date().toISOString(),
    });

    const agentUrl = `http://${state.agent.serviceFQDN}:3002`;
    console.log(`[${id}] sendPrompt: Sending to agent at ${agentUrl}`);

    // Accumulate assistant content for final persistence
    let assistantContent = "";

    try {
      // Make HTTP request to agent
      console.log(`[${id}] sendPrompt: Fetching ${agentUrl}/prompt`);
      const response = await fetch(`${agentUrl}/prompt`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ sessionId: id, text }),
      });

      console.log(`[${id}] sendPrompt: Agent response status=${response.status}`);
      if (!response.ok) {
        throw new Error(`Agent returned ${response.status}`);
      }

      // Stream SSE response
      const reader = response.body?.getReader();
      if (!reader) {
        throw new Error("No response body");
      }
      console.log(`[${id}] sendPrompt: Starting SSE stream`);

      const decoder = new TextDecoder();
      let buffer = "";

      const emit = (message: ServerMessage) => {
        console.log(`[${id}] sendPrompt: Emitting ${message.type} to ${state.messageHandlers.size} handlers`);
        for (const handler of state.messageHandlers) {
          handler(message);
        }
      };

      const handleSseData = async (data: unknown) => {
        if (!data || typeof data !== "object") return;
        const payload = data as {
          type?: string;
          content?: unknown;
          partial?: unknown;
          event?: unknown;
          error?: unknown;
          message?: unknown;
        };

        switch (payload.type) {
          case "agent.message": {
            const content =
              typeof payload.content === "string" ? payload.content : "";
            const partial =
              typeof payload.partial === "boolean" ? payload.partial : undefined;

            if (partial) {
              // Accumulate partial content
              assistantContent += content;
            } else {
              // Full message - use as final content
              assistantContent = content;
            }

            emit({
              type: "agent.message",
              sessionId: id,
              content,
              ...(partial !== undefined ? { partial } : {}),
            });
            return;
          }
          case "agent.event": {
            if (payload.event) {
              const event = payload.event as AgentEvent;

              // Persist event to Redis
              const eventId = `evt_${crypto.randomUUID().slice(0, 12)}`;
              await this.storage.appendEvent(id, {
                id: eventId,
                sessionId: id,
                event,
                timestamp: new Date().toISOString(),
              });

              emit({
                type: "agent.event",
                sessionId: id,
                event,
              });
            }
            return;
          }
          case "error":
          case "agent.error": {
            const errorMessage =
              typeof payload.error === "string"
                ? payload.error
                : typeof payload.message === "string"
                  ? payload.message
                  : "Unknown error";
            emit({
              type: "agent.error",
              sessionId: id,
              error: errorMessage,
            });
            return;
          }
          case "agent.system":
          case "agent.result":
          case "start":
          case "done":
          default:
            return;
        }
      };

      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        buffer += decoder.decode(value, { stream: true });

        // Parse SSE events
        const lines = buffer.split("\n");
        buffer = lines.pop() || ""; // Keep incomplete line in buffer

        for (const line of lines) {
          if (!line.startsWith("data: ")) continue;
          const payload = line.slice(6);
          try {
            await handleSseData(JSON.parse(payload));
          } catch {
            emit({
              type: "agent.message",
              sessionId: id,
              content: payload,
            });
          }
        }
      }

      // Persist final assistant message
      if (assistantContent) {
        const assistantMessageId = `msg_${crypto.randomUUID().slice(0, 12)}`;
        await this.storage.appendMessage(id, {
          id: assistantMessageId,
          sessionId: id,
          role: "assistant",
          content: assistantContent,
          timestamp: new Date().toISOString(),
        });

        // Emit done with messageId
        emit({ type: "agent.done", sessionId: id });
      } else {
        emit({ type: "agent.done", sessionId: id });
      }

      state.session.status = "ready";
      await this.storage.updateSessionField(id, "status", "ready");
    } catch (error) {
      console.error(`[${id}] Prompt error:`, error);
      for (const handler of state.messageHandlers) {
        handler({
          type: "agent.error",
          sessionId: id,
          error: error instanceof Error ? error.message : "Unknown error",
        });
      }
      state.session.status = "ready";
      await this.storage.updateSessionField(id, "status", "ready");
      throw error;
    }
  }

  /**
   * Interrupt the current agent operation
   */
  async interrupt(id: string): Promise<void> {
    const state = this.sessions.get(id);
    if (!state?.agent?.connected || !state.agent.serviceFQDN) return;

    try {
      await fetch(`http://${state.agent.serviceFQDN}:3002/interrupt`, {
        method: "POST",
        signal: AbortSignal.timeout(5000),
      });
    } catch (error) {
      console.error(`[${id}] Interrupt error:`, error);
    }
  }

  subscribe(id: string, handler: MessageHandler): () => void {
    const state = this.sessions.get(id);
    if (!state) {
      throw new Error(`Session ${id} not found`);
    }

    state.messageHandlers.add(handler);
    return () => state.messageHandlers.delete(handler);
  }

  // === Sync Methods ===

  /**
   * Get all sessions with metadata for sync
   */
  async getAllSessionsWithMeta(): Promise<SessionWithMeta[]> {
    const sessions = await this.list();
    const results: SessionWithMeta[] = [];

    for (const session of sessions) {
      const messageCount = await this.storage.getMessageCount(session.id);
      const lastMessage = await this.storage.getLastMessage(session.id);

      results.push({
        ...session,
        messageCount,
        lastMessageId: lastMessage?.id,
      });
    }

    return results;
  }

  /**
   * Get session with full history for sync
   */
  async getSessionWithHistory(
    id: string,
    afterMessageId?: string
  ): Promise<{
    session: Session;
    messages: PersistedMessage[];
    events: PersistedEvent[];
  } | null> {
    const state = this.sessions.get(id);
    if (!state) return null;

    const messages = await this.storage.getMessages(id, afterMessageId);
    const events = await this.storage.getEvents(id);

    return {
      session: state.session,
      messages,
      events,
    };
  }

  private emitSession(state: SessionState, message: ServerMessage): void {
    for (const handler of state.messageHandlers) {
      handler(message);
    }
  }

  private async startSessionCreation(
    id: string,
    request: SessionCreateRequest
  ): Promise<void> {
    const state = this.sessions.get(id);
    if (!state) return;

    try {
      // Create sandbox via Kubernetes
      console.log(`[${id}] Creating sandbox...`);
      await runtime.createSandbox({
        sessionId: id,
        cpus: config.defaultCpus,
        memoryMB: config.defaultMemoryMB,
        env: request.repo ? { GIT_REPO: request.repo } : undefined,
      });

      // Wait for sandbox to be ready
      console.log(`[${id}] Waiting for sandbox to be ready...`);
      const serviceFQDN = await runtime.waitForReady(id, 120000);
      if (!serviceFQDN) {
        state.session.status = "error";
        await this.storage.updateSessionField(id, "status", "error");
        this.emitSession(state, {
          type: "session.updated",
          session: state.session,
        });
        this.emitSession(state, {
          type: "session.error",
          id,
          error: "Sandbox failed to become ready",
        });
        return;
      }

      state.agent = {
        serviceFQDN,
        connected: true,
        close: () => {
          if (state.agent) state.agent.connected = false;
        },
      };

      state.session.status = "running";
      state.session.lastActiveAt = new Date().toISOString();

      // Persist status change
      await this.storage.updateSessionFields(id, {
        status: "running",
        lastActiveAt: state.session.lastActiveAt,
      });

      console.log(`[${id}] Session running at ${serviceFQDN}`);
      this.emitSession(state, {
        type: "session.updated",
        session: state.session,
      });
    } catch (error) {
      state.session.status = "error";
      await this.storage.updateSessionField(id, "status", "error");
      this.emitSession(state, {
        type: "session.updated",
        session: state.session,
      });
      this.emitSession(state, {
        type: "session.error",
        id,
        error:
          error instanceof Error ? error.message : "Failed to create session",
      });
      throw error;
    }
  }

  // TODO: Implement snapshots using Kubernetes VolumeSnapshots
  // async createSnapshot(id: string, name: string): Promise<void> { }
  // async restoreSnapshot(id: string, name: string): Promise<void> { }
  // async listSnapshots(id: string): Promise<string[]> { }
}

File: apps/control-plane/internal/session/manager.go
package session

import (
	"context"
	"fmt"
	"log/slog"
	"sync"
	"time"

	"github.com/angristan/netclode/apps/control-plane/internal/config"
	"github.com/angristan/netclode/apps/control-plane/internal/k8s"
	"github.com/angristan/netclode/apps/control-plane/internal/protocol"
	"github.com/angristan/netclode/apps/control-plane/internal/storage"
	"github.com/google/uuid"
)

const (
	sandboxReadyTimeout = 120 * time.Second
)

// Manager handles session lifecycle and agent communication.
type Manager struct {
	storage storage.Storage
	k8s     k8s.Runtime
	config  *config.Config

	sessions map[string]*SessionState
	mu       sync.RWMutex
}

// NewManager creates a new session manager.
func NewManager(store storage.Storage, k8sRuntime k8s.Runtime, cfg *config.Config) *Manager {
	return &Manager{
		storage:  store,
		k8s:      k8sRuntime,
		config:   cfg,
		sessions: make(map[string]*SessionState),
	}
}

// Initialize loads sessions from storage and reconciles with K8s state.
func (m *Manager) Initialize(ctx context.Context) error {
	slog.Info("Initializing session manager")

	// Load all sessions from storage
	sessions, err := m.storage.GetAllSessions(ctx)
	if err != nil {
		return fmt.Errorf("load sessions: %w", err)
	}

	slog.Info("Loaded sessions from storage", "count", len(sessions))

	// Load into memory with channel-based state
	for _, s := range sessions {
		m.sessions[s.ID] = NewSessionState(s)
	}

	// Reconcile with K8s
	sandboxes, err := m.k8s.ListSandboxes(ctx)
	if err != nil {
		slog.Warn("Failed to list sandboxes, skipping reconciliation", "error", err)
		return nil
	}

	sandboxMap := make(map[string]k8s.SandboxInfo)
	for _, sb := range sandboxes {
		sandboxMap[sb.SessionID] = sb
	}

	// Update session statuses based on K8s state
	for id, state := range m.sessions {
		sb, exists := sandboxMap[id]
		if !exists {
			// No sandbox exists
			if state.Session.Status == protocol.StatusRunning || state.Session.Status == protocol.StatusCreating {
				slog.Info("Session has no sandbox, marking as paused", "sessionID", id)
				state.Session.Status = protocol.StatusPaused
				_ = m.storage.UpdateSessionStatus(ctx, id, protocol.StatusPaused)
			}
		} else if sb.Ready {
			state.ServiceFQDN = sb.ServiceFQDN
			if state.Session.Status == protocol.StatusCreating {
				state.Session.Status = protocol.StatusReady
				_ = m.storage.UpdateSessionStatus(ctx, id, protocol.StatusReady)
			}
		}
	}

	slog.Info("Session manager initialized", "sessions", len(m.sessions), "sandboxes", len(sandboxes))
	return nil
}

// Close closes all session states.
func (m *Manager) Close() {
	m.mu.Lock()
	defer m.mu.Unlock()

	for _, state := range m.sessions {
		state.Close()
	}
}

// generateID creates a new session ID.
func generateID() string {
	return uuid.NewString()[:12]
}

// Create creates a new session.
func (m *Manager) Create(ctx context.Context, name string, repo *string) (*protocol.Session, error) {
	id := generateID()
	now := time.Now().UTC().Format(time.RFC3339)

	if name == "" {
		name = "Session " + id[:6]
	}

	session := &protocol.Session{
		ID:           id,
		Name:         name,
		Status:       protocol.StatusCreating,
		Repo:         repo,
		CreatedAt:    now,
		LastActiveAt: now,
	}

	// Save to storage
	if err := m.storage.SaveSession(ctx, session); err != nil {
		return nil, fmt.Errorf("save session: %w", err)
	}

	// Add to memory with channel-based state
	state := NewSessionState(session)
	m.mu.Lock()
	m.sessions[id] = state
	m.mu.Unlock()

	// Start sandbox creation in background
	go m.createSandbox(context.Background(), id, repo)

	return session, nil
}

func (m *Manager) createSandbox(ctx context.Context, sessionID string, repo *string) {
	env := map[string]string{
		"SESSION_ID":        sessionID,
		"ANTHROPIC_API_KEY": m.config.AnthropicAPIKey,
	}
	if repo != nil && *repo != "" {
		env["GIT_REPO"] = *repo
	}

	// Create sandbox
	if err := m.k8s.CreateSandbox(ctx, sessionID, env); err != nil {
		slog.Error("Failed to create sandbox", "sessionID", sessionID, "error", err)
		m.updateSessionStatus(ctx, sessionID, protocol.StatusError)
		m.emit(sessionID, protocol.NewSessionError(sessionID, err.Error()))
		return
	}

	// Wait for ready using informer-based watching
	fqdn, err := m.k8s.WaitForReady(ctx, sessionID, sandboxReadyTimeout)
	if err != nil {
		slog.Error("Sandbox failed to become ready", "sessionID", sessionID, "error", err)
		m.updateSessionStatus(ctx, sessionID, protocol.StatusError)
		m.emit(sessionID, protocol.NewSessionError(sessionID, err.Error()))
		return
	}

	// Update state
	m.mu.Lock()
	if state, ok := m.sessions[sessionID]; ok {
		state.ServiceFQDN = fqdn
		state.Session.Status = protocol.StatusRunning
	}
	m.mu.Unlock()

	if err := m.storage.UpdateSessionStatus(ctx, sessionID, protocol.StatusRunning); err != nil {
		slog.Error("Failed to update session status", "sessionID", sessionID, "error", err)
	}

	// Emit update
	if session := m.getSession(sessionID); session != nil {
		m.emit(sessionID, protocol.NewSessionUpdated(session))
	}

	slog.Info("Session created and running", "sessionID", sessionID, "fqdn", fqdn)
}

func (m *Manager) updateSessionStatus(ctx context.Context, sessionID string, status protocol.SessionStatus) {
	m.mu.Lock()
	if state, ok := m.sessions[sessionID]; ok {
		state.Session.Status = status
	}
	m.mu.Unlock()

	_ = m.storage.UpdateSessionStatus(ctx, sessionID, status)

	if session := m.getSession(sessionID); session != nil {
		m.emit(sessionID, protocol.NewSessionUpdated(session))
	}
}

// waitForSandbox waits for an existing sandbox to become ready (used when sandbox already exists).
func (m *Manager) waitForSandbox(ctx context.Context, sessionID string) {
	fqdn, err := m.k8s.WaitForReady(ctx, sessionID, sandboxReadyTimeout)
	if err != nil {
		slog.Error("Sandbox failed to become ready", "sessionID", sessionID, "error", err)
		m.updateSessionStatus(ctx, sessionID, protocol.StatusError)
		m.emit(sessionID, protocol.NewSessionError(sessionID, err.Error()))
		return
	}

	// Update state
	m.mu.Lock()
	if state, ok := m.sessions[sessionID]; ok {
		state.ServiceFQDN = fqdn
		state.Session.Status = protocol.StatusRunning
	}
	m.mu.Unlock()

	if err := m.storage.UpdateSessionStatus(ctx, sessionID, protocol.StatusRunning); err != nil {
		slog.Error("Failed to update session status", "sessionID", sessionID, "error", err)
	}

	// Emit update
	if session := m.getSession(sessionID); session != nil {
		m.emit(sessionID, protocol.NewSessionUpdated(session))
	}

	slog.Info("Sandbox ready", "sessionID", sessionID, "fqdn", fqdn)
}

// Resume resumes a paused session.
func (m *Manager) Resume(ctx context.Context, id string) (*protocol.Session, error) {
	m.mu.Lock()
	state, ok := m.sessions[id]
	m.mu.Unlock()

	if !ok {
		return nil, fmt.Errorf("session %s not found", id)
	}

	// Check if sandbox exists and is ready
	status, err := m.k8s.GetStatus(ctx, id)
	if err != nil {
		return nil, fmt.Errorf("get sandbox status: %w", err)
	}

	if status.Exists && status.Ready && status.ServiceFQDN != "" {
		// Sandbox already running
		m.mu.Lock()
		state.ServiceFQDN = status.ServiceFQDN
		state.Session.Status = protocol.StatusRunning
		m.mu.Unlock()

		_ = m.storage.UpdateSessionStatus(ctx, id, protocol.StatusRunning)
		return state.Session, nil
	}

	if status.Exists {
		// Sandbox exists but not ready yet - just wait for it
		m.mu.Lock()
		state.Session.Status = protocol.StatusCreating
		m.mu.Unlock()

		_ = m.storage.UpdateSessionStatus(ctx, id, protocol.StatusCreating)

		// Wait for existing sandbox to become ready
		go m.waitForSandbox(context.Background(), id)

		return state.Session, nil
	}

	// No sandbox exists - create new one
	m.mu.Lock()
	state.Session.Status = protocol.StatusCreating
	m.mu.Unlock()

	_ = m.storage.UpdateSessionStatus(ctx, id, protocol.StatusCreating)

	// Start sandbox creation in background
	go m.createSandbox(context.Background(), id, state.Session.Repo)

	return state.Session, nil
}

// Pause pauses a running session.
func (m *Manager) Pause(ctx context.Context, id string) (*protocol.Session, error) {
	m.mu.Lock()
	state, ok := m.sessions[id]
	m.mu.Unlock()

	if !ok {
		return nil, fmt.Errorf("session %s not found", id)
	}

	// Delete sandbox (but keep PVC)
	if err := m.k8s.DeleteSandbox(ctx, id); err != nil {
		slog.Warn("Failed to delete sandbox", "sessionID", id, "error", err)
	}

	// Delete secret
	if err := m.k8s.DeleteSecret(ctx, id); err != nil {
		slog.Warn("Failed to delete secret", "sessionID", id, "error", err)
	}

	// Update state
	m.mu.Lock()
	state.ServiceFQDN = ""
	state.Session.Status = protocol.StatusPaused
	m.mu.Unlock()

	_ = m.storage.UpdateSessionStatus(ctx, id, protocol.StatusPaused)

	return state.Session, nil
}

// Delete deletes a session and all its resources.
func (m *Manager) Delete(ctx context.Context, id string) error {
	// Close the session state (stops broadcast loop)
	m.mu.Lock()
	if state, ok := m.sessions[id]; ok {
		state.Close()
	}
	delete(m.sessions, id)
	m.mu.Unlock()

	// Delete K8s resources
	_ = m.k8s.DeleteSandbox(ctx, id)
	_ = m.k8s.DeleteSecret(ctx, id)
	_ = m.k8s.DeletePVC(ctx, id)

	// Delete from storage
	if err := m.storage.DeleteSession(ctx, id); err != nil {
		return fmt.Errorf("delete session from storage: %w", err)
	}

	slog.Info("Session deleted", "sessionID", id)
	return nil
}

// List returns all sessions, reconciling with K8s state.
func (m *Manager) List(ctx context.Context) ([]protocol.Session, error) {
	// Reconcile with K8s
	sandboxes, err := m.k8s.ListSandboxes(ctx)
	if err != nil {
		slog.Warn("Failed to list sandboxes", "error", err)
	}

	sandboxMap := make(map[string]k8s.SandboxInfo)
	for _, sb := range sandboxes {
		sandboxMap[sb.SessionID] = sb
	}

	m.mu.Lock()
	defer m.mu.Unlock()

	sessions := make([]protocol.Session, 0, len(m.sessions))
	for id, state := range m.sessions {
		sb, exists := sandboxMap[id]
		if !exists && (state.Session.Status == protocol.StatusRunning || state.Session.Status == protocol.StatusCreating) {
			state.Session.Status = protocol.StatusPaused
			_ = m.storage.UpdateSessionStatus(ctx, id, protocol.StatusPaused)
		} else if exists && sb.Ready {
			state.ServiceFQDN = sb.ServiceFQDN
		}
		sessions = append(sessions, *state.Session)
	}

	return sessions, nil
}

// Get returns a session by ID.
func (m *Manager) Get(ctx context.Context, id string) (*protocol.Session, error) {
	m.mu.RLock()
	state, ok := m.sessions[id]
	m.mu.RUnlock()

	if !ok {
		return nil, fmt.Errorf("session %s not found", id)
	}

	return state.Session, nil
}

// GetWithHistory returns a session with its message and event history.
func (m *Manager) GetWithHistory(ctx context.Context, id string, messageLimit int) (*protocol.Session, []protocol.PersistedMessage, []protocol.PersistedEvent, bool, error) {
	session, err := m.Get(ctx, id)
	if err != nil {
		return nil, nil, nil, false, err
	}

	messages, err := m.storage.GetMessages(ctx, id, nil)
	if err != nil {
		return nil, nil, nil, false, fmt.Errorf("get messages: %w", err)
	}

	events, err := m.storage.GetEvents(ctx, id, m.config.MaxEventsPerSession)
	if err != nil {
		return nil, nil, nil, false, fmt.Errorf("get events: %w", err)
	}

	// Convert to value slices
	msgSlice := make([]protocol.PersistedMessage, len(messages))
	for i, msg := range messages {
		msgSlice[i] = *msg
	}

	evtSlice := make([]protocol.PersistedEvent, len(events))
	for i, e := range events {
		evtSlice[i] = *e
	}

	hasMore := len(messages) >= messageLimit

	return session, msgSlice, evtSlice, hasMore, nil
}

// GetAllWithMeta returns all sessions with metadata.
func (m *Manager) GetAllWithMeta(ctx context.Context) ([]protocol.SessionWithMeta, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	result := make([]protocol.SessionWithMeta, 0, len(m.sessions))
	for id, state := range m.sessions {
		meta := protocol.SessionWithMeta{Session: *state.Session}

		count, err := m.storage.GetMessageCount(ctx, id)
		if err == nil {
			meta.MessageCount = &count
		}

		lastMsg, err := m.storage.GetLastMessage(ctx, id)
		if err == nil && lastMsg != nil {
			meta.LastMessageID = &lastMsg.ID
		}

		result = append(result, meta)
	}

	return result, nil
}

func (m *Manager) getSession(id string) *protocol.Session {
	m.mu.RLock()
	defer m.mu.RUnlock()

	if state, ok := m.sessions[id]; ok {
		return state.Session
	}
	return nil
}

func (m *Manager) getState(id string) *SessionState {
	m.mu.RLock()
	defer m.mu.RUnlock()

	return m.sessions[id]
}

// Subscribe returns a subscriber for a session.
// The subscriber's Send channel receives messages broadcast to the session.
func (m *Manager) Subscribe(id string) (*Subscriber, error) {
	state := m.getState(id)
	if state == nil {
		return nil, fmt.Errorf("session %s not found", id)
	}
	return state.Subscribe(), nil
}

// Unsubscribe removes a subscriber from a session.
func (m *Manager) Unsubscribe(id string, sub *Subscriber) {
	state := m.getState(id)
	if state != nil {
		state.Unsubscribe(sub)
	}
}

// emit sends a message to all subscribers of a session.
func (m *Manager) emit(id string, msg protocol.ServerMessage) {
	state := m.getState(id)
	if state != nil {
		state.Broadcast(msg)
	}
}
